### PR TITLE
[Runtime] Make Swift errors bridged to NSError and cloned round-trippable

### PR DESCRIPTION
**Pull request by DougGregor**
When a Swift error is bridged to NSError, it maintains enough information
to round-trip back to Swift. However, if the NSError is then cloned—e.g.,
a new NSError is created with the same domain/code/user-info dictionary,
the identity of the Swift type is lost.

When we initially form the user-info dictionary for a Swift-defined
error that is not explicitly bridgeable back from an NSError instance,
add a new key to the user-info dictionary containing a wrapped copy of the
original error. When trying to bridge from NSError, look for that key in
the user-info dictionary and cast to its value.

Fixes rdar://problem/29882046.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: stdlib/public/SDK/Foundation/NSError.swift [hunk 1]
```diff
@@ -217,6 +217,25 @@ public extension Error {
 internal let _errorDomainUserInfoProviderQueue = DispatchQueue(
   label: "SwiftFoundation._errorDomainUserInfoProviderQueue")
 
+/// Wraps up a Swift error instance so it can be stashed in the user-info
+/// dictionary of the NSError.
+internal class _SwiftOriginalErrorBox<T>: NSObject {
```
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSError.swift]: Does it need to inherit from NSObject?

**Review by jrose-apple — COMMENTED** [review]
Sad but clever. I wonder if we should just always use this instead of making NSError subclasses, so that we don't have two copies of the actual Error floating around.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**